File: src/bfs/bfs.hs
data Node = Node { value :: Int, adj :: [Node] } deriving Eq

instance Show Node where
    show node = show $ value node

-- source node의 집합
type Graph = [Node]

bfs :: Graph -> [Node]
bfs graph = [y | x <- graph, y <- bfs_node [x] [x] ]
    where bfs_node [] _ = []
          bfs_node queue visited = queue ++ bfs_node [y | x <- queue, y <- adj x, y `notElem` visited] (visited ++ queue)

File: src/bfs/bfs.py
import queue;

class Node:
    def __init__(self, value, adj):
        self.value = value
        self.adj = adj
        self.visited = False

        
def bfs(graph):
    q = queue.Queue()
    for source in graph:
        q.put(source)
        source.visited = True
        while q.qsize() != 0:
            top = q.get();
            print(top.value)
            
            for e in top.adj:
                if not e.visited:
                    e.visited = True
                    q.put(e)